datastore_entity/src/lib.rs

```diff
@@ -123,26 +118,33 @@ where
     query.limit = Some(1);
     req.query = Some(query);
 
-    let resp: RunQueryResponse = projects.run_query(req, connection.get_project_name()).execute().map_err(
-        |_e: google_datastore1::Error| -> String { "Failed to fetch entity by prop".to_string() },
-    )?;
+    let resp: RunQueryResponse = projects
+        .run_query(req, connection.get_project_name())
+        .execute()?;
 
     match resp.batch {
-        Some(batch) => {
-            // TODO - Validate more results -> ther shall not be any more results!
-            let found = batch.entity_results.unwrap(); // TODO - check before unwrap
-            for f in found {
-                // TODO - Validate legth instead of just returning the first element
-                if let Some(entity) = f.entity {
-                    let props = DatastoreProperties::from_map(entity.properties.unwrap());
-                    let result = DatastoreEntity::from(entity.key, props);
-
-                    return Ok(result);
+        Some(batch) => { 
+            if let Some(mut found) = batch.entity_results {
+                match found.len() {
+                    0 => Err(DatastoreClientError::NotFound)?,
+                    1 => {
+                        if let Some(entity) = found.remove(0).entity {
+                            let entity_properties = entity.properties.ok_or(DatastoreParseError::NoProperties)?;
+                            let props = DatastoreProperties::from_map(entity_properties);
+                            let result = DatastoreEntity::from(entity.key, props);
+    
+                            Ok(result)
+                        } else {
+                            Err(DatastoreClientError::NotFound)?
```

Comment:
Why does it need "?"?

Reply:
It is because the function return a DatastorersError, DatastorersError can contain a DatastoreClientError as one of its enum values (`DatastoreClientError`). the `#[from]` macro from thiserror implements `From<DatastoreClientError> for DatastorersError` so the `?`will wrap `DatastoreClientError::NotFound` into a DatastorersError.